<html>
<link type="text/css" rel="stylesheet" href="webcgh-help.css">
<body>
<h1 align="center" class="docTitle">WebGenome Help</h1>
<h2 class="heading1">Table of Contents</h2>
<a href="#overview">1 - WebGenome Version 3.0 Overview</a>
<br>
<a href="#cart">2 - The User's Workspace</a>
<br>
<a href="#plots">3 - Plots</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#plotTypes">3.1 - Plot Types</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#genomicPlots">3.1.1 - Genomic Plot Types</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#scatterPlot">3.1.1.1 - Scatter Plot</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#ideogramPlot">3.1.1.2 - Ideogram Plot</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#annotationPlot">3.1.1.3 - Annotation Plot</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#barPlot">3.1.2 - Bar Plot</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#plotParams">3.2 - Plotting Parameters</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#commonPlotParameters">3.2.1 - Common Plot Paremeters</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#genomicPlotParameters">3.2.2 - Genomic Plot Parameters</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#commonGenomicPlotParameters">3.2.2.1 - Common Genomic Plot Parameters</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#scatterPlotParameters">3.2.2.2 - Scatter Plot Parameters</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#ideogramPlotParameters">3.2.2.3 - Ideogram Plot Parameters</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#barPlotParameters">3.2.3 - Bar Plot Parameters</a>
<br>
<a href="#operations">4 - Analytic Operations</a>
<br>
<a href="#operations">5 - Appendix</a>
<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="#loh">5.1 - LOH Measurements</a>
<br>
<p>
<br>
</p>
<p>
<br>
</p>
<a name="overview"></a><span class="heading1">1 - WebGenome Version 3.0 Overview</span>
<br>
<p>
		
<p>
			WebGenome is a web-based application for
			creating genomics plots of array-based data
			and genomic annotation.  Version 2.3 provides a client
			plotting service to external applications, such as
			Rembrandt.  Users must enter webGenome through redirection
			from another application; the only way to load data is
			by selecting it in the referring client application.
		</p>
		
		
<p>
			Data types supported in this version include
			copy number (i.e., aCGH),
			<a href="#loh">loss of heterozygosity (LOH)</a>,
			and gene expression.  Two basic types of plots
			can be generated:
		</p>
		
		
<ul>
<a name=""></a>
<li>
<a href="#scatterPlot">Scatter Plots</a> -
					
				Plots DNA copy number measurements
				across the genome, chromosome, or chromosomal interval.
			</li>
<a name=""></a>
<li>
<a href="#ideogramPlot">Ideogram Plot</a> -
					
				Shows chromosomal amplifications
				and deletions in relation to cytogenetic chromosome ideograms.
			</li>
</ul>
		
		
<p>
			The underlying data model used by webGenome is a very simplied version of
			the MAGE-OM.  The most important concept for the use of the system
			is that individual array data sets, called <em>bioassays</em> are aggregated
			into experiments.
		</p>
	
</p>
<a name="cart"></a><span class="heading1">2 - The User's Workspace</span>
<br>
<p>
		
<p>
			The user's workspace holds accumulated raw data sets loaded by the client
			application, statistically processed data sets derived from
			raw or other processed data, and generated plots.  For
			each experiment, the list of nested bioassays is also shown.
			Users may change the names of experiments and bioassays in this
			screen.  Names do not have to be unique and can be empty
			if desired.  Users may delete experiments and plots.
		</p>
		
		
<p>
			Each bioassay has an associated color.  These colors are
			used in scatter plots to distinguish between bioassays.
			Colors are initialized by the system.  Users may change
			colors, and all future plots will use the new colors.  However,
			previously-generated plots in the cart will retain the original
			colors.
		</p>
		
		
<p>
			From the workspace, users may select one or more experiments
			and generate new plots or perform analytic operations.  New plots
			and statistically processed experiments are saved back into the
			cart.
		</p>
	
</p>
<a name="plots"></a><span class="heading1">3 - Plots</span>
<br>
<p>
	In this section we describe plotting functionality including the
	types of plots that can be generated and plotting parameters that
	can be set by the user.  Functionality to create new plots
	is accessible from the workspace screen.
	
	
		
		
		
		
	</p>
<a name="plotTypes"></a><span class="heading2">3.1 - Plot Types</span>
<br>
<p>
			WebGenome 3.0 includes three genomic plot types
			as well as a bar plot for selected reporters.
			
			
			
			
			
			
			
		</p>
<a name="genomicPlots"></a><span class="heading3">3.1.1 - Genomic Plot Types</span>
<br>
<p>
		
				
				
				
				
				
				
				
				
				
			
</p>
<a name="scatterPlot"></a><span class="heading4">3.1.1.1 - Scatter Plot</span>
<br>
<p>
					
<p>
					Scatter plots display array reporter values (e.g. copy number)
					against physical genome location.  Users can plot data points,
					regression lines, and confidence intervals in these plots.
					If data are loss of
					<a href="#loh">heterozygosity (LOH)</a>, then the
					endpoints of LOH regions are inferred and shown as
					thick horizontal lines.  The Y-coordinates of these lines
					are equal to the average probability of
					LOH weighted
					by the chromosomal distance between adjacent reporters.
					</p>
					
					
<p>
					Scatter plots provide basic interactivity.  Clicking
					on a data point causes all data points and lines from
					the bioassay to be highlighted.  Also the name of the
					selected bioassay will be outlined in the plot legend.
					Moving the
					mouse across the plot will cause a tool tip to show
					the nearest reporter.  If available, nearby genes will
					also be displayed along with arbitrary annotations that
					may be available through the client application.
					</p>
				
</p>
<a name="ideogramPlot"></a><span class="heading4">3.1.1.2 - Ideogram Plot</span>
<br>
<p>
					This plot shows color-coded data values in relation to
					chromosome ideograms.  The color coding goes from
					green to blue to red, which indicates data values that
					are low, medium, and high, respectively.  Moving the mouse
					over ideograms causes a tool tip to display the cytoband
					and physical chromosomal location under the pointer.
					Moving the mouse over data tracks causes a tool tip
					to indicate the physical chromosomal interval and data value
					under the pointer.  If available, nearby genes will
					also be displayed along with arbitrary annotations that
					may be available through the client application.
				</p>
<a name="annotationPlot"></a><span class="heading4">3.1.1.3 - Annotation Plot</span>
<br>
<p>
				Annotation plots.
				</p>
<a name="barPlot"></a><span class="heading3">3.1.2 - Bar Plot</span>
<br>
<p>
				This plot graphs only selected reporters.  The measurements
				at each reporter are shown using bars.  Bars are grouped
				by reporter.  Within each reporter group, each
				measurement from a different bioassay is shown as a single
				bar.
			</p>
<a name="plotParams"></a><span class="heading2">3.2 - Plotting Parameters</span>
<br>
<p>
			Plotting parameters control how data will be presented graphically.
			Some parameters that are common to all plot types,
			such as genome intervals.  Some parameters specific to one
			type of plot.  After the user elects to create a new plot
			from the workspace, a form showing parameters for
			the selected plot type is displayed.  The user may
			later change these parameter settings by clicking on
			a link above displayed plots.  Descriptions of the
			various parameters are given here.

			
			
			
			
			
		</p>
<a name="commonPlotParameters"></a><span class="heading3">3.2.1 - Common Plot Paremeters</span>
<br>
<p>
				These parameters are common to all plots.
				<ul>
<a name="param-plotName"></a>
<li>
<span class="propName">Plot name</span> -
					
						This parameter enables the user to name the plot.
						This name need not be unique and can be changed
						at any time with no harm.
					</li>
<a name="param-genomeIntervals"></a>
<li>
<span class="propName">Genome intervals</span> -
					
						This parameter constrains the genome interval that will
						be displayed in plots.  Genome segments should be specified in the
						following format: CHROMOSOME[:FROM-TO][,CHROMOSOME[:FROM-TO]]*
						
						<p>
<b>Examples</b>
</p>
						
						
<p>
							3:40000-50000<br>
							3,5:140000-160000<br>
							1,2<br>
							1:5000-7000,2:30000-100000
						</p>
					
</li>
<a name="param-numPlotPerRow"></a>
<li>
<span class="propName">Number of plot per row</span> -
					
						When the user specifies multiple genome intervals,
						the intervals will be displayed in different plots,
						which will be laid out in rows.  This parameter
						specifies how many plots will be in a row.
					</li>
</ul>
			
</p>
<a name="genomicPlotParameters"></a><span class="heading3">3.2.2 - Genomic Plot Parameters</span>
<br>
<p>
			
				
			
				
				
				
			
</p>
<a name="commonGenomicPlotParameters"></a><span class="heading4">3.2.2.1 - Common Genomic Plot Parameters</span>
<br>
<p>
					The parameters in this section are common to all genomic plots.
					<ul>
<a name="param-showReporterNames"></a>
<li>
<span class="propName">Show reporter names</span> -
					
							If this property is checked, names of reporters
							near mouse will be displayed
							in the mouseover text.
						</li>
<a name="param-showAnnotation"></a>
<li>
<span class="propName">Show reporter annotation</span> -
					
							If this property is checked, then annotation associated
							with the reporter under the mouse will be displayed
							in the mouseover text.
						</li>
<a name="param-showGenes"></a>
<li>
<span class="propName">Show gene names</span> -
					
							If this property is checked, then names of genes
							near the location of the under the mouse pointer
							will be displayed
							in the mouseover text.
						</li>
</ul>
				
</p>
<a name="scatterPlotParameters"></a><span class="heading4">3.2.2.2 - Scatter Plot Parameters</span>
<br>
<p>
					The following parameters are specific to scatter plots.
					<ul>
<a name="param-lohThreshold"></a>
<li>
<span class="propName">LOH probability threshold</span> -
					
							<a href="#loh">Loss of heterozygosity (LOH)</a>
							data may be provided
							by the client application
							as probabilities.  This parameter is a threshold,
							above which probabilities will be interpred as
							indicative of LOH.
						</li>
<a name="param-interpolateLohEndpoints"></a>
<li>
<span class="propName">Interpoloate LOH endpoints</span> -
					
							This parameter is only applicable if the data are
							<a href="#loh">LOH(loss of heterozygosity)</a>.
							Scatter plots of
							LOH indicate LOH regions as thick horizontal
							lines.  If for two adjacent reporters one is indicating LOH
							and the other is not, the location where the
							transition occurs cannot be known.  If this parameter
							is checked, the system will interpolate the
							LOH endpoint using the slope of a straight line
							between data points corresponding to the two reporters.
							If unchecked, the endpoint will be at the LOH reporter.
						</li>
<a name="param-drawRawLohProbabilities"></a>
<li>
<span class="propName">Include raw LOH probabilities</span> -
					
							This parameter is only applicable if the data are
							<a href="#loh">LOH(loss of heterozygosity)</a>.
							Scatter plots of
							LOH indicate LOH regions as thick horizontal
							lines.  If this parameter is checked, then the underlying
							raw LOH probabilites are also plotted.
						</li>
<a name="param-minY"></a>
<li>
<span class="propName">Minimum plot value</span> -
					
							This sets the minimum value on the scatter plot Y-axis.
						</li>
<a name="param-maxY"></a>
<li>
<span class="propName">Maximum plot value</span> -
					
							This sets the maximum value on the scatter plot Y-axis.
						</li>
<a name="param-width"></a>
<li>
<span class="propName">Plot width in pixels</span> -
					
							This sets the width of each individual scatter plot
							in pixels.  If multiple genome intervals are selected,
							then this propery applies to each of the corresponding
							plots.
						</li>
<a name="param-height"></a>
<li>
<span class="propName">Plot height in pixels</span> -
					
							This sets the height of each individual scatter plot
							in pixels.  If multiple genome intervals are selected,
							then this propery applies to each of the corresponding
							plots.
						</li>
<a name="param-drawHorizGridLines"></a>
<li>
<span class="propName">Draw horizontal grid lines</span> -
					
							If checked, horizontal grid lines emenating from
							the major tic marks on the Y-axis will be drawn across
							scatter plot.  
						</li>
<a name="param-drawVertGridLines"></a>
<li>
<span class="propName">Draw vertical grid lines</span> -
					
							If checked, vertical grid lines emenating from
							the major tic marks on the X-axis will be drawn across
							scatter plot. 
						</li>
<a name="param-drawPoints"></a>
<li>
<span class="propName">Draw points</span> -
					
							If checked, indivdual data points will be painted for
							each datum.
						</li>
<a name="param-drawErrorBars"></a>
<li>
<span class="propName">Draw error bars if plotting means</span> -
					
							The averaging analytic operation also calculates
							standard error values.  When plotting averages in
							a scatter plot, if this is checked then error
							bars will be painted for each data point.
						</li>
<a name="param-interpolationType"></a>
<li>
<span class="propName">Type of interpolation between data points</span> -
					
							Enables the user to specify if and how data points
							will be connected.  Options include none, straight
							lines, steps, or splines.
						</li>
</ul>
				
</p>
<a name="ideogramPlotParameters"></a><span class="heading4">3.2.2.3 - Ideogram Plot Parameters</span>
<br>
<p>
					The following parameters are specific to ideogram plots.
					<ul>
<a name="param-ideogramSize"></a>
<li>
<span class="propName">Ideogram size</span> -
					
							This enables the user to set the relative height
							of chromosome ideograms.
						</li>
<a name="param-ideogramThickness"></a>
<li>
<span class="propName">Ideogram width</span> -
					
							This specifies the width of chromosome ideograms
							in pixels.
						</li>
<a name="param-trackWidth"></a>
<li>
<span class="propName">Data track width</span> -
					
							This specifies the width of data tracks in pixels.
						</li>
<a name="param-minMask"></a>
<li>
<span class="propName">Minimum data mask value</span> -
					
							Background data values within a
							range (e.g., -0.2 and 0.3) can be masked
							so that only values of significant
							magnitude are colored.  This parameter is the
							the lower end of this range.
						</li>
<a name="param-maxMask"></a>
<li>
<span class="propName">Maximum data mask value</span> -
					
							Background data values within a
							range (e.g., -0.2 and 0.3) can be masked
							so that only values of significant
							magnitude are colored.  This parameter is the
							the upper end of this range.
						</li>
<a name="param-minSaturation"></a>
<li>
<span class="propName">Minimum saturation value</span> -
					
							Data tracks color code measured values (e.g. copy number)
							from green to blue to red.  These colors correspond to
							low, medium, and high values, respectively.  The saturation
							values indicate which values are mapped to pure green
							or red.  Values below and above the saturation points
							are mapped to pure green and red, respectively.
							This parameter is for the lower saturation value.
						</li>
<a name="param-maxSaturation"></a>
<li>
<span class="propName">Maximum saturation value</span> -
					
							Data tracks color code measured values (e.g. copy number)
							from green to blue to red.  These colors correspond to
							low, medium, and high values, respectively.  The saturation
							values indicate which values are mapped to pure green
							or red.  Values below and above the saturation points
							are mapped to pure green and red, respectively.
							This parameter is for the upper saturation value.
						</li>
</ul>
				
</p>
<a name="barPlotParameters"></a><span class="heading3">3.2.3 - Bar Plot Parameters</span>
<br>
<p>
				The following parameters are specific to bar plots.
				
				<ul>
<a name="param-rowHeight"></a>
<li>
<span class="propName">Row height</span> -
					
						The bar plot groups bars into rows and columns by
						reporter.  Each cell contains all measurement
						from a single reporter.  (Each measurement will
						come from a different bioassay.)  This parameter
						specifies how high each row of columns is
						in pixels.
					</li>
<a name="param-barWidth"></a>
<li>
<span class="propName">Bar width</span> -
					
						This parameter determines the width of each
						individual bar in pixels.
					</li>
</ul>
			
</p>
<a name="operations"></a><span class="heading1">4 - Analytic Operations</span>
<br>
<p>
		In this section we describe the analytic operation that can
		be performed on data in the user's workspace.  Once a user
		has selected one or more experiments, they may elect to
		perform an analytic operation on these experiments.
		The resulting processed data will be deposited back in the
		user's workspace.  Analytic operations may be performed on
		all raw experiments (i.e., experiments loaded from the client
		application directory) and most processed data.  Processed
		experiments resulting from the
		<a href="#op-7">minimum common alterations</a> operation
		cannot be further processed because individual data points
		are dropped.
		
		<ul>
<a name="op-1"></a>
<li>
<span class="propName">Bioconductor aCGH smoother</span> -
					
				This operation invokes the aCGH hidden Markov model-based
				method from the Bioconductor software package to
				smooth data.
			</li>
<a name="op-2"></a>
<li>
<span class="propName">Average</span> -
					
				This operation performs chromosome position-wise averaging
				of all data points within the same experiment.
			</li>
<a name="op-3"></a>
<li>
<span class="propName">Range based filter</span> -
					
				This operation filters out all data with values
				between two endpoints and is
				used primarily for removing outliers.
				The upper and lower endpoints
				may be INFINITY or -INFINITY, respectively.  This operation
				includes the following parameters:
				<ul>
<a name="opParam-lowerBound"></a>
<li>
<span class="propName">Lower bound</span> -
					
						Lower filter endpoint
					</li>
<a name="opParam-upperBound"></a>
<li>
<span class="propName">Upper bound</span> -
					
						Upper filter endpoint
					</li>
</ul>
			
</li>
<a name="op-4"></a>
<li>
<span class="propName">Simple bioassay-based normalization</span> -
					
				This operation subtracts the bioassay specific mean or median
				from each observation. The user specifies the use of the mean
				or median as the normalization factor. The choice of median
				based normalization allows the user to lessen the effect
				of outliers.  This type of normalization is useful for making
				spot intensities comparable across arrays 
				as it removes an overall array effect; each array will be 
				"centered" on zero (or something 
				close to it if the median is used).
			</li>
<a name="op-5"></a>
<li>
<span class="propName">Simple experiment-based normalization</span> -
					
				This procedure subtracts the array specific mean or median from
				each observation then adds the experiment level mean or median
				back to each observation. The user specifies 
				the use of the mean or median as the normalization factor.
				The choice of median based 
				normalization allows the user to lessen the effect of outliers.
				This type of normalization 
				is useful for making spot intensities comparable across arrays 
				while preserving any overall 
				difference due to treatment effect. This will center each array 
				on the difference between 
				it's own array mean/median and the experiment wide mean/median. 
			</li>
<a name="op-6"></a>
<li>
<span class="propName">Sliding window smoother</span> -
					
				This operation smooths data using a "sliding window" approach. 
				The user
				specified a <em>Window size</em>.  The value of probe <em>i</em>
				is equal to the average value of probes <em>i - w/2</em> through 
				<em>i + w/2</em>, where <em>w</em> is the size of the window.
			</li>
<a name="op-7"></a>
<li>
<span class="propName">Minimum common alterations</span> -
					
				This operation identifies the set of common alteration across
				bioassays.  These alterations are defined by a
				<em>threshold</em> and a <em>minPercent</em> property.
				The <em>threshold</em> is used for scoring alterations.
				For instance, if the threshold is set to 0.5, then all 
				<a href="#loh">LOH</a> probability values greater than
				this value will be
				considered LOH.  The <em>minPercent</em> propety defines
				the minimum percentage of bioassays that must contain
				an alteration at a particular location for the alteration
				to be considered common.
			</li>
<a name="op-8"></a>
<li>
<span class="propName">Linear regression</span> -
					
				This operation peforms data smoothing using linear regression.
			</li>
<a name="op-9"></a>
<li>
<span class="propName">Loess regression</span> -
					
				This operation peforms data smoothing using Loess regression.
			</li>
</ul>
	
</p>
<a name="operations"></a><span class="heading1">5 - Appendix</span>
<br>
<p>
		
	
</p>
<a name="loh"></a><span class="heading2">5.1 - LOH Measurements</span>
<br>
<p>
		
		
<p>
			Loss of heterozygosity (LOH) is detected by measuring
			the amount of DNA from each
			allele from tumor cells and comparing this with the amounts
			found from normal tissue in the same individual or a reference sample.
			A skewed distribution in the tumor cells, compared with those
			found in the normal tissue, indicates an allelic
			imbalance. This imbalance can result from either loss of an allele
			(true LOH) or allelic amplification. Some earlier articles reporting
			LOH may, for technical reasons, have measured allelic amplification
			rather than loss of an allele.
		</p>
		
		
<p>
			LOH measurements are reported as the
			-log10 of the probability of continuous genomic region of
			homozygote calls.
			Thus, the larger the LOH value is, the more likely the value
			indicates a true LOH
			(or allelic amplification). The following formulae are
			used to compute LOH values:
		</p>
		
		
<img src="loh_formula.png" align="center" border="1">

		</p>
</body>
</html>
